oak_functions/loader/src/server.rs

```diff
@@ -353,6 +443,98 @@ fn check_export_function_signature(
     }
 }
 
+pub async fn apply_policy<F, S>(
+    policy: Option<Policy>,
+    function: F,
+) -> anyhow::Result<Response<Body>>
+where
+    F: std::marker::Send + 'static + FnOnce() -> S,
+    S: std::future::Future<Output = anyhow::Result<Response<Body>>> + std::marker::Send,
+{
+    match policy {
+        None => function().await,
+        Some(policy) => {
+            // Use tokio::spawn to actually run the tasks in parallel, for more accurate measurement
+            // of time.
+            let duration = policy.constant_process_time_millis;
+            let timer = tokio::spawn(async move {
+                tokio::time::sleep(Duration::from_millis(duration)).await;
+            });
+            let task = tokio::spawn(async move { function().await });
+
+            // Run the function and the timer tasks in parallel, and measure the time that it
+            // takes for the first task to complete.
+            let start = Instant::now();
+            let response = tokio::select! {
+                _ = timer => Err(anyhow::anyhow!("timeout!")),
+                rsp = task => rsp.unwrap()
+            };
+            let elapsed = start.elapsed().as_millis();
+
+            if elapsed < policy.constant_process_time_millis as u128 {
+                let diff = policy.constant_process_time_millis as u128 - elapsed;
+                tokio::time::sleep(std::time::Duration::from_millis(diff as u64)).await;
+                match response {
+                    Err(err) => policy.get_error_response(format!("{}", err).as_str()),
+                    Ok(rsp) => {
+                        // All responses from Oak-Functions have an additional
+                        // `oak-policy-padding` header.
+                        let size = estimate_size(&rsp)?;
+                        match size.cmp(&(policy.constant_response_size_bytes as i64)) {
+                            Ordering::Greater => policy.get_error_response("response too large"),
+                            Ordering::Less => policy.pad_response(rsp),
+                            Ordering::Equal => Ok(rsp),
+                        }
+                    }
+                }
+            } else {
+                policy.get_error_response("response not available")
+            }
+        }
+    }
+}
+
+/// Estimates the size of the given response.
+///
+/// Measures the size of the response when formatted according to
+/// [RFC-2616](https://tools.ietf.org/html/rfc2616#section-6).
+/// The only relevant parts of the response are the status code and the response body. We also add
+/// the `oak-policy-padding` header to all responses. This header name is therefore also included
+/// when estimating the size.
+fn estimate_size(response: &hyper::Response<Body>) -> anyhow::Result<i64> {
```

Comment:
This seems like it is trying to undo what the inner call has already packaged in an HTTP request.  How about making the Wasm call return the body of the request, and wrap it in an HTTP respose afterwards, in a straightforward way?

Reply:
Exactly, I think we can just return `Option<Vec<u8>>` for now? Since the only status is whether the wasm function actually wrote a response or not, right? Although I'm not sure there are other error conditons. But yes, something along those lines. This has the added bonus that the Wasm invocation logic is agnostic of HTTP, so it could be used for gRPC or any other wrapper protocol, which I think we may want to do at some point (and seems a good way of designing this layer of the API). WDYT?